pkg/utils/kubernetes/update.go

```diff
@@ -20,31 +20,50 @@ import (
 	"time"
 
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/util/wait"
 	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
 )
 
 // TryUpdate tries to apply the given transformation function onto the given object, and to update it afterwards.
 // It retries the update with an exponential backoff.
-func TryUpdate(ctx context.Context, backoff wait.Backoff, c client.Client, obj runtime.Object, transform func() error) error {
-	return tryUpdate(ctx, backoff, c, obj, c.Update, transform)
+func TryUpdate(ctx context.Context, backoff wait.Backoff, c client.Client, obj runtime.Object, scheme *runtime.Scheme, transform func() error) error {
+	return tryUpdate(ctx, backoff, c, obj, scheme, c.Update, transform)
 }
 
 // TryUpdateStatus tries to apply the given transformation function onto the given object, and to update its
 // status afterwards. It retries the status update with an exponential backoff.
-func TryUpdateStatus(ctx context.Context, backoff wait.Backoff, c client.Client, obj runtime.Object, transform func() error) error {
-	return tryUpdate(ctx, backoff, c, obj, c.Status().Update, transform)
+func TryUpdateStatus(ctx context.Context, backoff wait.Backoff, c client.Client, obj runtime.Object, scheme *runtime.Scheme, transform func() error) error {
+	return tryUpdate(ctx, backoff, c, obj, scheme, c.Status().Update, transform)
 }
 
-func tryUpdate(ctx context.Context, backoff wait.Backoff, c client.Client, obj runtime.Object, updateFunc func(context.Context, runtime.Object, ...client.UpdateOption) error, transform func() error) error {
+func tryUpdate(ctx context.Context, backoff wait.Backoff, c client.Client, obj runtime.Object, scheme *runtime.Scheme, updateFunc func(context.Context, runtime.Object, ...client.UpdateOption) error, transform func() error) error {
 	key, err := client.ObjectKeyFromObject(obj)
 	if err != nil {
 		return err
 	}
```

Comment:
What do you think about if we introduce something like this here:
```
resetCopy := obj.DeepCopyObject()

objGetter := func() runtime.Object {
  return resetCopy.DeepCopyObject()
}
```

The `objGetter` function can then be called in the `exponentialBackoff` part to get the object state before transformation in every iteration. I know that this is not the same as having a "fresh" object but it should still solve the issue explained in #2454.

Reply:
I am not sure I understand what you suggest. But assuming you meant to do obj = objGetter() we are going to lose the pointer to the object from the caller function and it wouldn't be affected, right?

Reply:
Yes you are right, I missed that one. The idea was to "reset" the object if a conflict occurred.

```
reflect.ValueOf(obj).Set(reflect.ValueOf(resetCopy))
```

Reply:
I tried it like you proposed and it seems that it is working fine. @timebertt  are you ok with such solution

Reply:
Nice, thanks for trying it out 👍

Reply:
Yes, I'm fine with it. Seems like a good solution, thanks 👍 
I would still try to go for using Patch requests wherever possible, but that is a separate refactoring, which we don't have to do right now. :)